features/steps/thens/alignment.py

```diff
@@ -352,3 +359,67 @@ def step_impl(context, inst, activation_phrase):
+
+
+@gherkin_ifc.step('Each segment must have geometric continuity in {continuity_type}')
+def step_impl(context, inst, continuity_type):
+    """
+    Assess geometric continuity between alignment segments for ALS016, ALS017, and ALS018
+    """
+    # Ref: https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcTransitionCode.htm
+    position_transition_codes = ["CONTINUOUS", "CONTSAMEGRADIENT", "CONTSAMEGRADIENTSAMECURVATURE"]
+    tangency_transition_codes = ["CONTSAMEGRADIENT", "CONTSAMEGRADIENTSAMECURVATURE"]
+
+    length_unit_scale_factor = ifcopenshell.util.unit.calculate_unit_scale(
+        ifc_file=context.model,
+        unit_type="LENGTHUNIT"
+    )
+    for pair in inst:
+        pair_continuity_was_calculated = False
+        previous, current = pair
+        entity_contexts = ifc.recurrently_get_entity_attr(context, current, 'IfcRepresentation', 'ContextOfItems')
+        precision = ifc.get_precision_from_contexts(entity_contexts)
+        precision_str = str(precision)
+        parts = precision_str.lower().split("e")
+        mantissa = parts[0]
+        sig_figs = len(mantissa)
+        if "." in mantissa:
+            sig_figs -= 1
+
+        exponent_str = parts[-1]
+        display_sig_figs = int(abs(float(exponent_str)) + sig_figs + 1)  # add one more digit for rounding confirmation
+
+        match continuity_type.lower():
+            case "position":
+                expected = precision
+                if current.Transition in position_transition_codes:
+                    observed = geometry.alignment_segment_positional_difference(
+                        length_unit_scale_factor,
+                        previous,
+                        current
+                    )
+                    pair_continuity_was_calculated = True
+
+            case "tangency":
+                if current.Transition in tangency_transition_codes:
+                    expected = math.atan2(precision, float(current.SegmentLength.wrappedValue) * 1000)
+                    observed = geometry.alignment_segment_angular_difference(
+                        length_unit_scale_factor,
+                        previous,
+                        current
+                    )
+                    pair_continuity_was_calculated = True
+
+        if pair_continuity_was_calculated and (abs(observed) > expected):
+            expected_msg = "{:.{width}e}".format(expected, width=display_sig_figs)
+            observed_msg = "{:.{width}e}".format(observed, width=display_sig_figs)
+
+            # add a suffix so that we keep the number formatting
+            # otherwise the values will be coerced back to floats
+            expected_msg += f" (max deviation in {continuity_type.lower()})"
+            observed_msg += f" (calculated deviation in {continuity_type.lower()})"
+            yield ValidationOutcome(
+                inst=current,
+                expected=expected_msg,
+                observed=observed_msg,
```

Comment:
I know this is annoying, but I'd really urge us to do formatting in the frontend to allow for internationalization and querying the db.

e.g `{'value': observed, 'ndigits': ..., 'context': ...}`

Reply:
Good point.  I have revised this.